dvc/command/dag.py

```diff
@@ -30,80 +29,91 @@ def _show_dot(G):
+
+        def _relabel(out):
+            return str(out)
 
-    if target:
-        H = get_pipeline(get_pipelines(G), target)
-        if not full:
-            descendants = nx.descendants(G, target)
-            descendants.add(target)
-            H.remove_nodes_from(set(G.nodes()) - descendants)
     else:
-        H = G
+        G = repo.graph
 
-    if outs:
-        G = nx.DiGraph()
-        for stage in H.nodes:
-            G.add_nodes_from(stage.outs)
+        def _relabel(stage):
+            return stage.addressing
 
-        for from_stage, to_stage in nx.edge_dfs(H):
-            G.add_edges_from(
-                [
-                    (from_out, to_out)
-                    for from_out in from_stage.outs
-                    for to_out in to_stage.outs
-                ]
-            )
-        H = G
+    return nx.relabel_nodes(G, _relabel, copy=True)
 
-    def _relabel(node):
-        from dvc.stage import Stage
 
-        return node.addressing if isinstance(node, Stage) else str(node)
+def _filter(G, targets, full):
+    import networkx as nx
 
-    return nx.relabel_nodes(H, _relabel, copy=False)
+    if not targets:
+        return G
+
+    H = G.copy()
+    if not full:
+        descendants = set()
+        for target in targets:
+            descendants.update(nx.descendants(G, target))
+            descendants.add(target)
+        H.remove_nodes_from(set(G.nodes()) - descendants)
+
+    undirected = H.to_undirected()
+    connected = set()
+    for target in targets:
+        connected.update(nx.node_connected_component(undirected, target))
+
+    H.remove_nodes_from(set(H.nodes()) - connected)
+
+    return H
+
+
+def _build(repo, target=None, full=False, outs=False):
+    targets = _collect_targets(repo, target, outs)
+    G = _transform(repo, outs)
+    return _filter(G, targets, full)
 
 
 class CmdDAG(CmdBase):
     def run(self):
-        try:
-            target = None
-            if self.args.target:
-                stages = self.repo.collect(self.args.target)
-                if len(stages) > 1:
-                    logger.error(
-                        f"'{self.args.target}' contains more than one stage "
-                        "{stages}, please specify one stage"
-                    )
-                    return 1
-                target = stages[0]
-
-            G = _build(
-                self.repo.graph,
-                target=target,
-                full=self.args.full,
-                outs=self.args.outs,
-            )
-
-            if self.args.dot:
-                logger.info(_show_dot(G))
-            else:
-                from dvc.utils.pager import pager
-
-                pager(_show_ascii(G))
-
-            return 0
-        except DvcException:
-            msg = "failed to show "
-            if self.args.target:
-                msg += f"a pipeline for '{target}'"
-            else:
-                msg += "pipelines"
-            logger.exception(msg)
-            return 1
```

Comment:
Unrelated :slightly_frowning_face: , but we are gradually going away from KO errors like this in other places, so did it here too.